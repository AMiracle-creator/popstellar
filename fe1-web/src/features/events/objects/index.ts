export * from './Event';
export * from './LaoEventBuilder';
