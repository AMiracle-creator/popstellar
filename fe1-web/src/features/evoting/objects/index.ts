export * from './Configuration';
export * from './Election';
export * from './Feature';
