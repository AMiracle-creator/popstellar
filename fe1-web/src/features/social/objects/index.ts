export * from './Chirp';
export * from './Reaction';
