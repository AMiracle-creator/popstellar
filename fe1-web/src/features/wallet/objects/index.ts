export * from './Seed';
export { generateToken, getCurrentPopTokenFromStore } from './Token';
