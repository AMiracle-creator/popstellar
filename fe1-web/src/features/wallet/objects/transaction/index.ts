export * from './DigitalCashHelper';
export * from './Transaction';
export * from './TransactionInput';
export * from './TransactionOutput';
