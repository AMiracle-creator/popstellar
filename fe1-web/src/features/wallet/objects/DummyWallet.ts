import testKeyPair from 'test_data/keypair.json';

import { EventTags, Hash, PublicKey, Timestamp } from 'core/objects';
import { dispatch } from 'core/redux';
import { addEvent, removeEvent } from 'features/events/reducer';
import { connectToLao, disconnectFromLao, removeLao } from 'features/lao/reducer';
import { RollCall, RollCallStatus } from 'features/rollCall/objects';
import { addRollCall, removeRollCall } from 'features/rollCall/reducer';

import { Lao, LaoState } from '../../lao/objects';
import { generateToken } from './Token';

// region Dummy values definition
const mockPublicKey = testKeyPair.publicKey;

const org = new PublicKey(mockPublicKey);

const mockLaoName = 'MyLao';
const mockLaoCreationTime = new Timestamp(1600000000);
const mockLaoIdHash: Hash = Hash.fromStringArray(
  org.toString(),
  mockLaoCreationTime.toString(),
  mockLaoName,
);
const mockLaoId: string = mockLaoIdHash.toString();

const mockLaoState: LaoState = {
  id: mockLaoId,
  name: mockLaoName,
  creation: mockLaoCreationTime.valueOf(),
  last_modified: mockLaoCreationTime.valueOf(),
  organizer: org.valueOf(),
  witnesses: [],
  server_addresses: [],
};
const mockLao = Lao.fromState(mockLaoState);

// MOCK ROLL CALL
const mockRCName = 'myRollCall';
const mockRCLocation = 'location';
const mockRCTimestampStart = new Timestamp(1620255600);
const mockRCTimestampEnd = new Timestamp(1620357600);
const mockRCAttendees = ['attendee1', 'attendee2'];

const mockRCIdHash = Hash.fromStringArray(
  EventTags.ROLL_CALL,
  mockLaoId,
  mockRCTimestampStart.toString(),
  mockRCName,
);

const mockRollCallState: any = {
  id: mockRCIdHash.valueOf(),
  eventType: RollCall.EVENT_TYPE,
  start: mockRCTimestampStart.valueOf(),
  name: mockRCName,
  location: mockRCLocation,
  creation: mockRCTimestampStart.valueOf(),
  proposedStart: mockRCTimestampStart.valueOf(),
  proposedEnd: mockRCTimestampEnd.valueOf(),
  status: RollCallStatus.CLOSED,
  attendees: mockRCAttendees,
};
// endregion

const createRollCall = (id: string, name: string, mockAttendees: string[]) => {
  return RollCall.fromState({
    ...mockRollCallState,
    id: id,
    name: name,
    attendees: [...mockRollCallState.attendees, ...mockAttendees],
  });
};

const mockRCName0: string = 'mock0';
const mockRCName1: string = 'mock1';

const hashMock0 = Hash.fromStringArray(
  EventTags.ROLL_CALL,
  mockLao.id.valueOf(),
  mockRollCallState.start.valueOf(),
  mockRCName0,
);
const hashMock1 = Hash.fromStringArray(
  EventTags.ROLL_CALL,
  mockLao.id.valueOf(),
  mockRollCallState.start.valueOf(),
  mockRCName1,
);

/*
 * Generates a mock state with some mock popTokens
 */
export async function createDummyWalletState() {
  const tokenMockRC0 = await generateToken(mockLao.id, hashMock0);
  const tokenMockRC1 = await generateToken(mockLao.id, hashMock1);
  const mockRollCall0 = createRollCall(hashMock0.valueOf(), mockRCName0, [
    tokenMockRC0.publicKey.valueOf(),
  ]);
  const mockRollCall1 = createRollCall(hashMock1.valueOf(), mockRCName1, [
    tokenMockRC1.publicKey.valueOf(),
  ]);
  dispatch(connectToLao(mockLao.toState()));
  dispatch(
    addEvent(mockLao.id, {
      eventType: RollCall.EVENT_TYPE,
      id: mockRollCall0.id.valueOf(),
      start: mockRollCall0.proposedStart.valueOf(),
      end: mockRollCall0.proposedEnd.valueOf(),
    }),
  );
  dispatch(addRollCall(mockRollCall0.toState()));

  dispatch(
    addEvent(mockLao.id, {
      eventType: RollCall.EVENT_TYPE,
      id: mockRollCall1.id.valueOf(),
      start: mockRollCall1.proposedStart.valueOf(),
      end: mockRollCall1.proposedEnd.valueOf(),
    }),
  );
  dispatch(addRollCall(mockRollCall1.toState()));
  console.debug('Dispatched mock events');
}
/*
 * Clears the mock state
 */
export function clearDummyWalletState() {
  dispatch(removeEvent(mockLao.id, hashMock0));
  dispatch(removeEvent(mockLao.id, hashMock1));
  dispatch(removeRollCall(hashMock0));
  dispatch(removeRollCall(hashMock1));
  dispatch(disconnectFromLao());
  dispatch(removeLao(mockLao.id));
}
