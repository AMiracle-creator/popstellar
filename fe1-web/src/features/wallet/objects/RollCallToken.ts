import { Hash, KeyPairState, PopToken } from 'core/objects';

export interface RollCallTokenState {
  token: KeyPairState;
  laoId: string;
  rollCallId: string;
  rollCallName: string;
}
/**
 * A Roll Call Token object, defined by a Pop token, its lao id hash and its Roll Call id hash
 */
export class RollCallToken {
  public readonly token: PopToken;

  public readonly laoId: Hash;

  public readonly rollCallId: Hash;

  public readonly rollCallName: string;

  constructor(obj: Partial<RollCallToken>) {
    if (obj === undefined || obj === null) {
      throw new Error(
        'Error encountered while creating a RollCallToken object: undefined/null parameters',
      );
    }

    if (obj.token === undefined) {
      throw new Error("Undefined 'token' when creating 'RollCallToken'");
    }
    if (obj.laoId === undefined) {
      throw new Error("Undefined 'laoId' when creating 'RollCallToken'");
    }
    if (obj.rollCallId === undefined) {
      throw new Error("Undefined 'rollCallId' when creating 'RollCallToken'");
    }
    if (obj.rollCallName === undefined) {
      throw new Error("Undefined 'rollCallName' when creating 'RollCallToken'");
    }

    this.token = obj.token;
    this.laoId = obj.laoId;
    this.rollCallId = obj.rollCallId;
    this.rollCallName = obj.rollCallName;
  }

  public static fromState(rct: RollCallTokenState): RollCallToken {
    return new RollCallToken({
      token: PopToken.fromState(rct.token),
      laoId: new Hash(rct.laoId),
      rollCallId: new Hash(rct.rollCallId),
      rollCallName: rct.rollCallName,
    });
  }

  public toState(): RollCallTokenState {
    return {
      token: this.token.toState(),
      laoId: this.laoId.valueOf(),
      rollCallId: this.rollCallId.valueOf(),
      rollCallName: this.rollCallName,
    };
  }
}
