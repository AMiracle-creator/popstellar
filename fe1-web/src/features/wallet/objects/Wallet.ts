import { Lao } from 'features/lao/objects';
import { RollCall } from 'features/rollCall/objects';

import { RollCallToken } from './RollCallToken';
import { generateToken } from './Token';

/**
 * Recovers all PoP tokens associated with this wallet.
 * @return Promise<RollCallToken[]>
 */
export async function recoverWalletRollCallTokens(
  rollCalls: Record<string, Record<string, RollCall>>,
  currentLao: Lao,
): Promise<RollCallToken[]> {
  // For all the roll calls of the current lao
  const tokens = Object.values(rollCalls[currentLao.id.valueOf()]).map((rc) => {
    // Generate the token corresponding to this roll call
    return generateToken(currentLao.id, rc.id).then((popToken) => {
      // If the token participated in the roll call, create a RollCallToken object
      if (rc.containsToken(popToken)) {
        return new RollCallToken({
          token: popToken,
          laoId: currentLao.id,
          rollCallId: rc.id,
          rollCallName: rc.name,
        });
      }
      return undefined;
    });
  });
  return (await Promise.all(tokens)).filter((rct) => rct !== undefined) as RollCallToken[];
}
