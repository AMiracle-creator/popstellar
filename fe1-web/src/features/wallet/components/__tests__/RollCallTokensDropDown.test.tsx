import { render } from '@testing-library/react-native';
import React from 'react';

import { mockKeyPair, mockLao } from '__tests__/utils';
import { Hash, PopToken } from 'core/objects';

import { RollCallToken } from '../../../../core/objects/RollCallToken';
import { RollCallTokensDropDown } from '../index';

const mockToken = PopToken.fromState(mockKeyPair.toState());
const mockArray = [
  new RollCallToken({
    token: mockToken,
    laoId: mockLao.id,
    rollCallId: new Hash('id'),
    rollCallName: 'rcname',
  }),
];

describe('RollCallTokensDropDown', () => {
  it('renders correctly', () => {
    const { toJSON } = render(
      <RollCallTokensDropDown
        rollCallTokens={mockArray}
        selectedTokenIndex={0}
        onIndexChange={() => {}}
      />,
    );
    expect(toJSON()).toMatchSnapshot();
  });
});
