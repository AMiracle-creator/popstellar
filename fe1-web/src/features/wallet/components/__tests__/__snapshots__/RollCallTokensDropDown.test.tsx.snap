// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`RoundIconButton renders correctly 1`] = `
<View
  style={
    Object {
      "fontSize": 20,
      "margin": "auto",
      "maxHeight": 40,
      "maxWidth": "90%",
      "textAlign": "center",
      "width": "fit-content",
    }
  }
>
  <RNCPicker
    items={
      Array [
        Object {
          "label": "J9fBzJV70Jk5c-i3277Uq4CmeL4t53WDfUghaK0HpeM=",
          "testID": undefined,
          "textColor": undefined,
          "value": 0,
        },
      ]
    }
    onChange={[Function]}
    selectedIndex={0}
    style={
      Array [
        Object {
          "height": 216,
        },
        undefined,
      ]
    }
  />
</View>
`;
